_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+
+The OpenID Connect core specification defines the following roles:
+
+1. End-User: Human participant
+2. Authorization Server: The server issuing access tokens to the client after successfully authenticating the resource owner and obtaining authorization.
+3. Client: An application making protected resource requests on behalf of the resource owner (the end-user) and with its authorization.
+
+From the OpenID Connect specification, the authentication using Authorization Code Flow has the following steps:
+
+{% img blog/spring-security-pkce/auth0-authorization-code.png alt:"Authorization Code Grant" width:"800" %}{: .center-image }
+
+1. Client prepares an authentication request and sends the request to the Authorization Server.
+2. Authorization Server prompts for the End-User authentication and obtains End-User consent/authorization.
+3. The End-User authenticates and gives consent
+4. Authorization Server sends the End-User back to the Client with an Authorization Code.
+5. Client requests a response using the Authorization Code at the Token Endpoint.
+6. Client receives a response that contains an ID Token and Access Token in the response body.
+
+The diagram is a simplified sequence of the Authorization Code Flow, where the User Agent (the browser) redirections are not shown.
+
+When calling the Token Endpoint, the client application must authenticate itself. The protocol provides several ways for [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication):
+
+- `client_secret_basic`: The client was assigned a secret (confidential client) and authenticates using HTTP basic authentication scheme.
+- `client_secret_post`: The client was assigned a secret and authenticates by including the client credentials in the request body.
+- `client_secret_jwt`: The client was assigned a secret and authenticates using a JWT signed with part of the secret as a shared key.
+- `private_key_jwt`: The client has registered a public key and authenticates using a JWT signed with the private key.
+- `none`: The client does not authenticate itself at the Token Endpoint, because it is a public client.
+
+As browser and mobile applications _cannot_ hold credentials securely and therefore cannot identify themselves using a client secret, [PKCE](https://www.rfc-editor.org/rfc/rfc7636) was created for extending the OAuth 2.0 Authorization Code Flow, adding a dynamically created cryptographically random key called "code verifier". This extension was created to mitigate the [authorization code interception attack](https://www.rfc-editor.org/rfc/rfc7636#section-1).
+
+The modified flow has the following steps:
+
+{% img blog/spring-security-pkce/auth0-authorization-code-pkce.png alt:"Authorization Code Grant" width:"850" %}{: .center-image }
```

Comment:
The diagrams still don't look uniform, can you recreate both diagrams in Powerpoint or Keynote or any other tool so that they look similar. If you want I can help with recreating them.

Reply:
I uploaded new versions. Would you please recreate one of them so I can copy the style?